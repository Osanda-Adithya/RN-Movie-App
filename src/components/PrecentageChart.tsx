import React, { FC } from "react";
import { View, StyleSheet, I18nManager, Text } from 'react-native'
import { ternary } from "../utils/Tools";

type PrecentageChartProps = {
    color: string,
    shadowColor: string,
    bgColor: string,
    radius: number,
    borderWidth: number,
    percent: number,
}

function percentToDegrees(percent: number) {
    return percent * 3.6
}

let direction = ternary(I18nManager.isRTL, 'right', 'left');
let opDirection = ternary(I18nManager.isRTL, 'Left', 'Right');

const PrecentageChart: FC<PrecentageChartProps> = (props) => {

    function computeDerivedState() {
        const percent = Math.max(Math.min(100, props.percent), 0)
        const needHalfCircle2 = percent > 50
        let halfCircle1Degree
        let halfCircle2Degree
        // degrees indicate the 'end' of the half circle, i.e. they span (degree - 180, degree)
        if (needHalfCircle2) {
            halfCircle1Degree = 180
            halfCircle2Degree = percentToDegrees(percent)
        } else {
            halfCircle1Degree = percentToDegrees(percent)
            halfCircle2Degree = 0
        }

        return {
            halfCircle1Degree,
            halfCircle2Degree,
            halfCircle2Styles: {
                // when the second half circle is not needed, we need it to cover
                // the negative degrees of the first circle
                backgroundColor: needHalfCircle2
                    ? props.color
                    : props.shadowColor,
            },
        }
    }

    function renderHalfCircle(rotateDegrees: number, halfCircleStyles?: any) {
        const { radius, color } = props
        return (
            <View
                style={[
                    styles.leftWrap,
                    {
                        width: radius,
                        height: radius * 2,
                    },
                ]}
            >
                <View
                    style={[
                        styles.halfCircle,
                        {
                            width: radius,
                            height: radius * 2,
                            borderRadius: radius,
                            overflow: 'hidden',
                            transform: [
                                { translateX: radius / 2 },
                                { rotate: `${rotateDegrees}deg` },
                                { translateX: -radius / 2 },
                            ],
                            backgroundColor: color,
                            ...halfCircleStyles,
                        },
                    ]}
                />
            </View>
        )
    }

    function renderInnerCircle() {
        const radiusMinusBorder = props.radius - props.borderWidth
        return (
            <View
                style={[
                    styles.innerCircle,
                    {
                        width: radiusMinusBorder * 2,
                        height: radiusMinusBorder * 2,
                        borderRadius: radiusMinusBorder,
                        backgroundColor: props.bgColor,
                    },
                ]}
            >
                <Text style={styles.precentage}>{Math.max(Math.min(100, props.percent), 0).toFixed(0)}%</Text>
            </View>
        )
    }

    const {
        halfCircle1Degree,
        halfCircle2Degree,
        halfCircle2Styles,
    } = computeDerivedState()

    return (
        <View
            style={[
                styles.outerCircle,
                {
                    width: props.radius * 2,
                    height: props.radius * 2,
                    borderRadius: props.radius,
                    backgroundColor: props.shadowColor
                },
            ]}
        >
            {renderHalfCircle(halfCircle1Degree)}
            {renderHalfCircle(halfCircle2Degree, halfCircle2Styles)}
            {renderInnerCircle()}
        </View>
    )

}

export default PrecentageChart;

const styles = StyleSheet.create({
    outerCircle: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    innerCircle: {
        overflow: 'hidden',
        justifyContent: 'center',
        alignItems: 'center',
    },
    leftWrap: {
        position: 'absolute',
        top: 0,
        [`${direction}`]: 0,
    },
    halfCircle: {
        position: 'absolute',
        top: 0,
        left: 0,
        [`borderTop${opDirection}Radius`]: 0,
        [`borderBottom${opDirection}Radius`]: 0,
    },
    precentage: {
        color: 'white',
        fontSize: 12
    }
})