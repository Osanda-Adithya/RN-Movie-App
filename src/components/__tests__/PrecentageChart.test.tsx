import { render } from "@testing-library/react-native";
import React from "react";
import PrecentageChart from "../PrecentageChart";

describe('render snapshot with less than 50', () => {
    it('render snapshot', () => {
        const container = render(
            <PrecentageChart
                percent={30}
                radius={50}
                borderWidth={8}
                color="#3399FF"
                shadowColor="#999"
                bgColor="#fff"
            />
        );
        expect(container.toJSON()).toMatchSnapshot();
    })

    it('render snapshot with more than 50', () => {
        const container = render(
            <PrecentageChart
                percent={80}
                radius={50}
                borderWidth={8}
                color="#3399FF"
                shadowColor="#999"
                bgColor="#fff"
            />
        );
        expect(container.toJSON()).toMatchSnapshot();
    })
})