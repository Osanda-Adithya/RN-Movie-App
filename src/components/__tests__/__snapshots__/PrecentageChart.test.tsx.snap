// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render snapshot with less than 50 render snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#999",
        "borderRadius": 50,
        "height": 100,
        "width": 100,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "height": 100,
          "width": 50,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomRightRadius": 0,
            "borderTopRightRadius": 0,
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          Object {
            "backgroundColor": "#3399FF",
            "borderRadius": 50,
            "height": 100,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 25,
              },
              Object {
                "rotate": "108deg",
              },
              Object {
                "translateX": -25,
              },
            ],
            "width": 50,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "height": 100,
          "width": 50,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomRightRadius": 0,
            "borderTopRightRadius": 0,
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          Object {
            "backgroundColor": "#999",
            "borderRadius": 50,
            "height": 100,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 25,
              },
              Object {
                "rotate": "0deg",
              },
              Object {
                "translateX": -25,
              },
            ],
            "width": 50,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 42,
          "height": 84,
          "width": 84,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 12,
        }
      }
    >
      30
      %
    </Text>
  </View>
</View>
`;

exports[`render snapshot with less than 50 render snapshot with more than 50 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#999",
        "borderRadius": 50,
        "height": 100,
        "width": 100,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "height": 100,
          "width": 50,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomRightRadius": 0,
            "borderTopRightRadius": 0,
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          Object {
            "backgroundColor": "#3399FF",
            "borderRadius": 50,
            "height": 100,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 25,
              },
              Object {
                "rotate": "180deg",
              },
              Object {
                "translateX": -25,
              },
            ],
            "width": 50,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "left": 0,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "height": 100,
          "width": 50,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomRightRadius": 0,
            "borderTopRightRadius": 0,
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          Object {
            "backgroundColor": "#3399FF",
            "borderRadius": 50,
            "height": 100,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 25,
              },
              Object {
                "rotate": "288deg",
              },
              Object {
                "translateX": -25,
              },
            ],
            "width": 50,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 42,
          "height": 84,
          "width": 84,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 12,
        }
      }
    >
      80
      %
    </Text>
  </View>
</View>
`;
